src/vm/op_contract.rs

```diff
@@ -246,28 +243,37 @@ impl InstructionSet for ContractOp {
                 );
             }
             ContractOp::CnG(state_type, reg) => {
-                regs.set_n(RegA::A16, *reg, context.global.get(state_type).map(|a| a.len_u16()));
+                regs.set_n(RegA::A8, *reg, context.global.get(state_type).map(|a| a.len_u16()));
             }
-            ContractOp::CnC(_state_type, _reg) => {
-                // TODO: implement global contract state
-                fail!()
+            ContractOp::CnC(state_type, reg) => {
+                regs.set_n(
+                    RegA::A16,
+                    *reg,
+                    context.owned_state.get(*state_type).map(|a| a.len_u16()),
+                );
             }
-            ContractOp::LdP(state_type, index, reg) => {
+            ContractOp::LdP(state_type, reg_32, reg) => {
+                let reg_32 = regs.get_n(RegA::A16, *reg_32);
```

Comment:
We must fail here and everywhere if the register with the index is not set:
```suggestion
                let Some(reg_32) = regs.get_n(RegA::A16, *reg_32) else { fail!() };
```